cpp/src/arrow/util/trie-test.cc

```diff
@@ -0,0 +1,283 @@
+    ASSERT_EQ(s.substr(1, 0), "");
+    ASSERT_EQ(s.substr(1, 2), "bc");
+    ASSERT_EQ(s.substr(4, 0), "");
+    ASSERT_EQ(s.substr(4, 1), "");
+  }
+}
+
+static std::vector<std::string> AllNulls() {
+  return {"#N/A",    "#N/A N/A", "#NA", "-1.#IND", "-1.#QNAN", "-NaN", "-nan", "1.#IND",
+          "1.#QNAN", "N/A",      "NA",  "NULL",    "NaN",      "n/a",  "nan",  "null"};
+}
+
+static void TestTrieContents(const Trie& trie, const std::vector<std::string>& entries) {
+  std::unordered_map<std::string, int32_t> control;
+  auto n_entries = static_cast<int32_t>(entries.size());
+
+  // Build control container
+  for (int32_t i = 0; i < n_entries; ++i) {
+    auto p = control.insert({entries[i], i});
+    ASSERT_TRUE(p.second);
+  }
+
+  // Check all existing entries in trie
+  for (int32_t i = 0; i < n_entries; ++i) {
+    ASSERT_EQ(i, trie.Find(entries[i])) << "for string '" << entries[i] << "'";
+  }
+
+  auto CheckNotExists = [&control, &trie](const std::string& s) {
+    auto p = control.find(s);
+    if (p == control.end()) {
+      ASSERT_EQ(-1, trie.Find(s)) << "for string '" << s << "'";
+    }
+  };
+
+  // Check potentially non-existing strings
+  CheckNotExists("");
+  CheckNotExists("X");
+  CheckNotExists("abcdefxxxxxxxxxxxxxxx");
+
+  // Check potentially non-existing variations of existing entries
+  for (const auto& e : entries) {
+    CheckNotExists(e + "X");
+    if (e.size() > 0) {
+      CheckNotExists(e.substr(0, 1));
+      auto prefix = e.substr(0, e.size() - 1);
+      CheckNotExists(prefix);
+      CheckNotExists(prefix + "X");
+      auto split_at = e.size() / 2;
+      CheckNotExists(e.substr(0, split_at) + 'x' + e.substr(split_at + 1));
+    }
+  }
+}
+
+static void TestTrieContents(const std::vector<std::string>& entries) {
+  TrieBuilder builder;
+  for (const auto& s : entries) {
+    ASSERT_OK(builder.Append(s));
+  }
+  const Trie trie = builder.Finish();
+  ASSERT_OK(trie.Validate());
+
+  TestTrieContents(trie, entries);
+}
+
+TEST(Trie, Empty) {
+  TrieBuilder builder;
+  const Trie trie = builder.Finish();
+  ASSERT_OK(trie.Validate());
+
+  ASSERT_EQ(-1, trie.Find(""));
+  ASSERT_EQ(-1, trie.Find("x"));
+}
+
+TEST(Trie, EmptyString) {
+  TrieBuilder builder;
+  ASSERT_OK(builder.Append(""));
+  const Trie trie = builder.Finish();
+  ASSERT_OK(trie.Validate());
+
+  ASSERT_EQ(0, trie.Find(""));
+  ASSERT_EQ(-1, trie.Find("x"));
+}
+
+TEST(Trie, Basics1) {
+  TestTrieContents({"abc", "de", "f"});
+  TestTrieContents({"abc", "de", "f", ""});
+}
+
+TEST(Trie, Basics2) {
+  TestTrieContents({"a", "abc", "abcd", "abcdef"});
+  TestTrieContents({"", "a", "abc", "abcd", "abcdef"});
+}
+
+TEST(Trie, Basics3) {
+  TestTrieContents({"abcd", "ab", "a"});
+  TestTrieContents({"abcd", "ab", "a", ""});
+}
+
+TEST(Trie, LongStrings) {
+  TestTrieContents({"abcdefghijklmnopqr", "abcdefghijklmnoprq", "defghijklmnopqrst"});
+  TestTrieContents({"abcdefghijklmnopqr", "abcdefghijklmnoprq", "abcde"});
```

Comment:
The long strings only create 2 nodes (right?), not sure if there's value in having to create more than 2 to ensure that the recursion is happening properly

Reply:
Hmm... right, our small strings should be 11 chars long.